include/internal/quic_record_rx.h

```diff
@@ -0,0 +1,507 @@
+ * and making the necessary calls to the TX side by detecting changes to the
+ * return value of ossl_qrx_get_key_epoch().
+ *
+ * The above states (NORMAL, UPDATING, COOLDOWN) can themselves be
+ * considered substates of the PROVISIONED state. Providing a secret to the QRX
+ * for an EL transitions from UNPROVISIONED, the initial state, to PROVISIONED
+ * (NORMAL). Dropping key material for an EL transitions from whatever the
+ * current substate of the PROVISIONED state is to the DISCARDED state, which is
+ * the terminal state.
+ *
+ * Note that non-1RTT ELs cannot undergo key update, therefore a non-1RT EL is
+ * always in the NORMAL substate if it is in the PROVISIONED state.
+ */
+
+/*
+ * Return the current RX key epoch for the 1-RTT encryption level. This is
+ * initially zero and is incremented by one for every Key Update successfully
+ * signalled by the peer. If the 1-RTT EL has not yet been provisioned or has
+ * been discarded, returns UINT64_MAX.
+ *
+ * A necessary implication of this API is that the least significant bit of the
+ * returned value corresponds to the currently expected Key Phase bit, though
+ * callers are not anticipated to have any need of this information.
+ *
+ * It is not possible for the returned value to overflow, as a QUIC connection
+ * cannot support more than 2**62 packet numbers, and a connection must be
+ * terminated if this limit is reached.
+ *
+ * The caller should use this function to detect when the key epoch has changed
+ * and use it to initiate a key update on the TX side.
+ *
+ * The value returned by this function increments specifically at the transition
+ * from the NORMAL to the UPDATING state discussed above.
+ */
+uint64_t ossl_qrx_get_key_epoch(OSSL_QRX *qrx);
+
+/*
+ * Relates to the 1-RTT encryption level. The caller should call this after the
+ * UPDATING state is reached, after a timeout to be determined by the caller.
+ *
+ * This transitions from the UPDATING state to the COOLDOWN state (if
+ * still in the UPDATING state). If normal is 1, then transitions from
+ * the COOLDOWN state to the NORMAL state. Both transitions can be performed at
+ * once if desired.
+ *
+ * If in the normal state, or if in the COOLDOWN state and normal is 0, this is
+ * a no-op and returns 1. Returns 0 if the 1-RTT EL has not been provisioned or
+ * has been dropped.
+ *
+ * It is essential that the caller call this within a few PTO intervals of a key
+ * update occurring (as detected by the caller in a call to
+ * ossl_qrx_key_get_key_epoch()), as otherwise the peer will not be able to
+ * perform a Key Update ever again.
+ */
+int ossl_qrx_key_update_timeout(OSSL_QRX *qrx, int normal);
+
+
+/*
+ * Key Expiration
+ * ==============
+ */
+
+/*
+ * Returns the number of seemingly forged packets which have been received by
+ * the QRX. If this value reaches the value returned by
+ * ossl_qrx_get_max_epoch_forged_pkt_count(), all further received encrypted
```

Comment:
There are two limits - confidentiality limit - that should limit the number of packets sent with a given key. This is TX side limit so should be handled on the TX record layer.

The integrity limit (i.e., the count of received packets that weren't successfully decrypted and authenticated) applies to the RX side and if I understand the RFC 9001 correctly it applies across all keys in the connection so it is not per epoch. There is also no way how to escape the integrity limit by key update - if the limit is reached `the endpoint MUST immediately close the connection with a connection error of type AEAD_LIMIT_REACHED and not process any more packets.`